yoto_api/YotoAPI.py

```diff
@@ -22,61 +24,236 @@
+        try:
+            data = {
+                "access_token": token.access_token,
+                "refresh_token": token.refresh_token,
+                "token_type": token.token_type,
+                "scope": token.scope,
+                "valid_until": token.valid_until.isoformat()
+                if token.valid_until
+                else None,
+            }
+
+            with open(self.token_file, "w") as f:
+                json.dump(data, f)
+
+            # Ensure file permissions are secure
+            self.token_file.chmod(0o600)
+
+        except Exception as e:
+            _LOGGER.error(f"Failed to save token: {e}")
+
+    def _is_valid(self, token: Token) -> bool:
+        if not token or not token.access_token or not token.valid_until:
+            return False
+        # 5 minute buffer
+        return token.valid_until > datetime.datetime.now(pytz.utc) + timedelta(
+            minutes=5
+        )
+
+    def clear_token(self) -> None:
+        if self.token_file.exists():
+            self.token_file.unlink()
+
+    def login(self, client_id: str) -> Token:
+        if not client_id:
+            raise ValueError("client_id is required")
 
-    # https://api.yoto.dev/#75c77d23-397f-47f9-b76c-ce3c647b11d5
-    def login(self, username: str, password: str, client_id: str) -> Token:
-        self.CLIENT_ID = client_id
-        url = f"{self.BASE_URL}/auth/token"
+        # Step 1: Get authorization details
+        auth_response = self._get_authorization(client_id)
+
+        # Step 2: Display user instructions
+        self._display_user_instructions(auth_response)
+
+        # Step 3: Poll for token
+        return self._poll_for_token(client_id, auth_response)
+
+    def _get_authorization(self, client_id: str) -> dict:
+        """Get authorization code and user instructions."""
         data = {
             "audience": self.BASE_URL,
             "client_id": client_id,
-            "grant_type": "password",
-            "password": password,
-            "username": username,
-            "scope": "openid email profile offline_access",
+            "scope": "offline_access",
         }
         headers = {"Content-Type": "application/x-www-form-urlencoded"}
 
-        response = requests.post(url, data=data, headers=headers)
-        data = response.json()
+        response = requests.post(self.AUTH_URL, data=data, headers=headers)
+
+        if not response.ok:
+            raise AuthenticationError(
+                f"Authorization failed: {response.status_code} {response.text}"
+            )
 
-        _LOGGER.debug(f"{DOMAIN} - Sign In Response {data.keys()}")
+        return response.json()
 
-        if response.status_code != 200:
-            if data["error"] == "invalid_grant":
-                raise AuthenticationError(data["error_description"])
-            else:
-                raise Exception(data["error_description"])
+    def _display_user_instructions(self, auth_result: dict) -> None:
```

Comment:
This won't work for home assistant since the print statements won't be displayed.  Would be logging.